### ISSUE TITLE
Refactor UI with a modern design system

### ISSUE THREAD

**Issue by zhuiks** [ezra-bible-app/ezra-bible-app#196]
UI can be refactored! 
Instead of jQuery-UI it might be one of the following well established frameworks:
- [Bootstrap 5](https://getbootstrap.com/docs/5.0)
- [Material](https://material.io/components?platform=web)

This will allow to have more responsive layout suitable for mobile devices

I know it's a lot of work but I think it should happen sometime in the future.

**Comment by tobias-klein** [comment]
These are the things that would have to be re-worked:

* Tabs
* Accordion widget
* Dialogs
* Module Install Assistant
* Verse selection mechanism
* Buttons

Most effort is probably the Tabs and the Module Install Assistant.

Technically this issue could also be handled incrementally, which would require that both libraries (JQuery UI plus the new one) are loaded in parallel until all rework is done.

Before we attempt this work I would like to see a clear list of benefits for the user and/or what this enables in the future.

I would estimate that this is probably at least a full week of work (full-time) or potentially even more.

How would you estimate the effort?

**Comment by zhuiks** [comment]
> Before we attempt this work I would like to see a clear list of benefits for the user and/or what this enables in the future.
- Benefits for the user:
  - Mobile ready! 
  - Modern attractive design 
  - Familiar interface / design solutions
  - Possibility for right-to-left (RTL) languages

- Benefits for devs and the project
  - simplified css. Most styles would come from the framework/theme. So when we are adding new component we don't need to copy/create styles for it. 
  - simplified and semantic html with established html5 and css3 technologies (i.e. flexbox, grid)
  - much more of the common ui task are already implemented within the framework (although it's true for the current jQuery-ui).
  - frontend code can be refactored to be even more component based, i.e. have logic (js) markup (hrml) and few specific styles (css) are in the same file/folder. If we decide sometime in the future to switch to something like react/vue - having such structure would be very beneficial. 
  - more design components to work with for future enhancements
  - more familiar code for new contributors

> Tabs

[Tabs](https://material.io/archive/guidelines/components/tabs.html)

> Accordion widget

[Expansion Panels](https://material.io/archive/guidelines/components/expansion-panels.html#)

> Dialogs

[Dialogs](https://material.io/archive/guidelines/components/dialogs.html)

> Module Install Assistant

[Steppers](https://material.io/archive/guidelines/components/steppers.html)

> Verse selection mechanism

That one has to be customized. But there is a [familiar pattern](https://material.io/archive/guidelines/patterns/selection.html#selection-item-selection)

> Buttons

[Buttons](https://material.io/archive/guidelines/components/buttons.html)

I can try to draw a prototype in [Figma](https://www.figma.com/@materialdesign). I think it would be beneficial for the future to experiment with UI.
I think we should keep the current color palette 

> Technically this issue could also be handled incrementally, which would require that both libraries (JQuery UI plus the new one) are loaded in parallel until all rework is done.

Yes definitely it should be done incrementally. Maybe it can be even a separate subproject. Maybe for a candidate for 1.0?

> How would you estimate the effort?

It's all depends on how deeply we want to refactor and how familiar dev is with the project code. For me it would be  >70 hours (2 full weeks). I assume a lot of small nuances might pop up.

**Comment by tobias-klein** [comment]
Thanks for the extensive analysis!

From my perspective, the biggest benefits (based on your list) are:
- Mobile ready. This would open the door for having a responsive design that scales from smartphone to tablet to desktop. I would not want to tune the current html/css design towards a smartphone. It works decently on tablet/desktop, but going smaller would not be sustainable with the current framework.
- Simplified and component-based CSS.
- More design components to work with for future enhancements.

In terms of an implementation strategy I could imagine the following:

Start with a new layout / CSS infrastructure from scratch, while trying to re-use existing controllers/components and make them work both with the old and the new layout for a transition period.

0) Decide on the new/modified technology stack.
1) Setup a rough framework with HTML/CSS and basic boot-up code from scratch. Optimize for mobile-first plus add CSS media query adjustments for tablet + desktop sizes. Test in a regular web-browser (with index.html + css + initial JS code).
2) Integrate bible browsing functionality based on existing backend + controllers.
3) Integrate menu bar and option handling.
4) Integrate module assistant.
5) Implement tab system based on existing controllers/components.
6) Integrate tag assignment. & tag browsing from existing components.
7) Integrate in-tab search from existing components.
8) Integrate dictionary functionality from existing components.

In terms of technology I think it's a good idea to introduce new responsive design based on modern CSS framework(s). On the other hand I would be a bit relucant regarding JS-frameworks like React, Vue, etc. and rather stick to Vanilla JS (primarily for performance reasons) with a similar approach to components as already used today.

If this can be finished rather quickly (I doubt that) this could happen on a separate development branch.
If it's a longer-term activity then this could happen in a separate repository that includes the current repo as a submodule and re-uses code that way.

Any further thoughts on implementation strategy?

**Comment by zhuiks** [comment]
I would agree with _separate repository_ (I can use my fork of Ezra Project) and _Vanilla JS_ (future proof)

I see three main goals for this subproject:
1. **Redesign** UI with mobile first in mind
2. Select and embrace design system with its **component library**
3. **Refactor** code along the way (unite html, css, js under each component, possibly implement state management solution for frontend)

In general I will also agree with your roadmap. But I would revised it after first new component implementation and add design prototype step:

0. Decide on the technology
1. Setup the technology
2. Refactor a small component (to understand the concept and try out the technology). Maybe **Fullscreen button**
3. Draw design prototype and propose it for open discussion. 
4. Re-evaluate roadmap
- Refactor Bible browsing
- Tags, notes
- Tabs, menu, etc...
- Module assistant
- ...



## 0. Requirements for modified technology stack:
- Good established **Design system**
As developers we want a "rich" component library so we can just add a few lines of code to use already developed components (tabs, accordion, dialog modals, popups, charts). But design system is much bigger than just a number of components. It has philosophy behind it and a guidance for possible new components.
- **Custom elements / web components**
It's already a [standard](https://developer.mozilla.org/en-US/docs/Web/Web_Components#specifications) so it will stay till all the browsers will die. It's a nice, framework agnostic way to hide implementation details. It's a learning curve, but it seems like it's future proof.
- **Popular**
I think we need to choose something well known by people vs personal preference. That way if somebody new comes into the project he/she might be already familiar with it.

A good list to choose from can be found [here](https://open-wc.org/guides/community/component-libraries/)

From my research I'd point out [Carbon](https://github.com/carbon-design-system/carbon-custom-elements) (IBM) and [Material](https://material-components.github.io/material-components-web-components/demos/index.html) (Google). They are very big.
**Carbon** seems to be more close and enterprise oriented, although it has nice clear documentation and separate section on data visualization. I can't find any examples other than ibm.com

**Material** is the most popular, familiar for end-users who have android devices, and very thought-trough system.

My personal preference is Material. What do you think? @tobias-klein @alerque

**Comment by alerque** [comment]
I'm going to bow out of this one, I'm kind of following discussions at a distance but I'm not very closely involved in web front end things these days and don't have a good feel for what the pros and cons of various frameworks are.

From a logistics standpoint stay away from Git submodules as a way of reaching into a code base to reuse bits. I use them every day and they serve a purpose, but this would not be a good use case for them. I would either work in a development branch here, or if you want to split out issues and discussion just start with a fork of this repository and work on top of that with a separate issue tracker. Personally I think just splitting it out here with appropriate issue tags or a project board and it's own branch will be better. If this refactoring does go live it should be at least on par with the current one and therefore replace it across the board so I don't see the point of working on it as a separate project.

**Comment by tobias-klein** [comment]
Since this issue is so large, I'm also turning this into a discussion for now.
Whenever we come up with "digestible" implementation proposals we can raise new issues for those. We need to take it step by step / incrementally and each increment could be a new issue.
